src/network/linkers_socket.cpp

```diff
@@ -187,19 +187,20 @@ void Linkers::Construct() {
   listener_->Listen(incoming_cnt);
   std::thread listen_thread(&Linkers::ListenThread, this, incoming_cnt);
   const int connect_fail_retry_cnt = 20;
-  const int connect_fail_delay_time = 10 * 1000;  // 10s
   // start connect
   for (auto it = need_connect.begin(); it != need_connect.end(); ++it) {
     int out_rank = it->first;
     // let smaller rank connect to larger rank
     if (out_rank > rank_) {
       TcpSocket cur_socket;
+      int connect_fail_delay_time = 200;  // 0.2s
       for (int i = 0; i < connect_fail_retry_cnt; ++i) {
         if (cur_socket.Connect(client_ips_[out_rank].c_str(), client_ports_[out_rank])) {
           break;
         } else {
           Log::Warning("Connecting to rank %d failed, waiting for %d milliseconds", out_rank, connect_fail_delay_time);
           std::this_thread::sleep_for(std::chrono::milliseconds(connect_fail_delay_time));
+          connect_fail_delay_time = static_cast<int>(connect_fail_delay_time * 1.5);
```

Comment:
@guolinke now that we're using this backoff strategy, I think `connect_fail_retry_cnt` should be reduced. Under the current state (on `master`) it would take 200s to determine no connectivity. Now it will take roughly 1329s to determine no connectivity (summing `0.2 * 1.5^n` for `n` in `0:19`). That means if we merge this PR, `LightGBM` will now takes 6x as long to diagnose a connectivity problem.

I like this new backoff strategy but think we should consider reducing `connect_fail_retry_cnt` to, say, 16 (which would mean it takes 262s to determine no connectivity). 

Would you consider that change? Or was the intent of this PR to ALSO wait longer to declare connection lost?

Also, per #2348 do we eventually want to make this retry logic something user-configurable?